pkg/brokerpak/fetch.go

```diff
@@ -25,7 +40,91 @@ func fetchArchive(src, dest string) error {
 		Src:           src,
 		Dst:           dest,
 		Mode:          getter.ClientModeFile,
-		Getters:       getter.Getters,
 		Decompressors: map[string]getter.Decompressor{},
 	}).Get()
 }
+
+// fetchBrokerpak downloads a local or remote brokerpak; brokerpaks can be
+// fetched remotely using the gs:// prefix which will load them from a
+// Cloud Storage bucket with the broker's credentials.
+// Relative paths are resolved relative to the executable.
+func fetchBrokerpak(src, dest string) error {
+	execWd := filepath.Dir(os.Args[0])
+	execDir, err := filepath.Abs(execWd)
+	if err != nil {
+		return fmt.Errorf("couldn't turn dir %q into abs path: %v", execWd, err)
+	}
+
+	getters := map[string]getter.Getter{}
+	for k, g := range getter.Getters {
+		getters[k] = g
+	}
+	getters["gs"] = &gsGetter{}
+
+	return (&getter.Client{
+		Src: src,
+		Dst: dest,
+		Pwd: execDir,
+
+		Mode:          getter.ClientModeFile,
+		Getters:       getters,
+		Decompressors: map[string]getter.Decompressor{},
+	}).Get()
+}
+
+// gsGetter is a go-getter that works on Cloud Storage using the broker's
+// service account. It's incomplete in that it doesn't support directories.
+type gsGetter struct{}
+
+// ClientMode is unsupported for gsGetter.
+func (g *gsGetter) ClientMode(u *url.URL) (getter.ClientMode, error) {
+	return getter.ClientModeInvalid, errors.New("mode is not supported for this client")
+}
+
+// Get clones a remote destination
+func (g *gsGetter) Get(dst string, u *url.URL) error {
+	return errors.New("getting directories is not supported for this client")
+}
+
+// GetFile downloads the give URL into the given path. The URL must
+// reference a single file. If possible, the Getter should check if
+// the remote end contains the same file and no-op this operation.
+func (g *gsGetter) GetFile(dst string, u *url.URL) error {
+	ctx, cancel := context.WithTimeout(context.Background(), 5*time.Minute)
+	defer cancel()
+
+	obj, err := g.objectAt(u)
+	if err != nil {
+		return err
+	}
+
+	src := stream.FromReadCloserError(obj.NewReader(ctx))
+	return stream.Copy(src, stream.ToFile(dst))
+}
+
+func (g *gsGetter) objectAt(u *url.URL) (*storage.ObjectHandle, error) {
+	client, err := g.client()
```

Comment:
This method seems to have two responsibilities: 1. initialize the GCS client 2. fetch the object. If you separated these concerns and had the client be passed in to the constructor of the getter.Client , it would improve orthogonality and allow for unit testing the logic of this module via mocking the client.

Reply:
I agree that the `objectAt` method should accept a `*storage.Client` parameter.

Reply:
Since the only logic here is to trim the leading slash, I think this is fine to skip on unit testing and reply on integration testing only. Since to unit test you'd have to create a implementation of the storage.Client interface, which I agree is a lot of work for simply passing through a `gs://` url